### ISSUE TITLE
improve type definition

### ISSUE THREAD

**Issue by 435352980** [sanusart/react-dropdown-select#76]
```ts
declare module "react-dropdown-select" {
  import React, { KeyboardEvent, CSSProperties, ChangeEvent } from "react";

  export interface SetStateFnArgs<T> {
    dropdown?: boolean;
    values?: T[];
    search?: string;
    selectBounds?: DOMRect | {};
    cursor?: number | null;
    activeCursorItem?: any;
  }

  export interface SelectState<T> {
    dropdown: boolean;
    values: T[];
    search: string;
    selectBounds: object;
    cursor: number;
  }

  export interface SelectMethods<T> {
    removeItem: (
      event: React.MouseEvent<HTMLElement, MouseEvent> | null,
      item: T,
      close: boolean
    ) => void;
    dropDown: (
      action: string,
      event?: React.MouseEvent<HTMLElement, MouseEvent> | null
    ) => void;
    addItem: (item: T) => void;
    setSearch: (event: ChangeEvent<HTMLInputElement>) => void;
    getInputSize: () => number;
    toggleSelectAll: () => void;
    clearAll: () => void;
    selectAll: (items?: T[]) => void;
    searchResults: () => T[];
    getSelectRef: () => HTMLDivElement;
    isSelected: (item: T) => boolean;
    getSelectBounds: () => {} | DOMRect;
    areAllSelected: () => boolean;
    handleKeyDown: (event: KeyboardEvent) => void;
    activeCursorItem: (activeCursorItem: any) => void;
    createNew: (searchText: string) => void;
    sortBy: () => T[];
    safeString: (input: string) => string;
  }

  export interface SelectRenderer<T> {
    props: SelectProps<T>;
    state: SelectState<T>;
    methods: SelectMethods<T>;
  }

  export interface SelectKeyDown<T> {
    event: KeyboardEvent;
    props: SelectProps<T>;
    state: SelectState<T>;
    methods: SelectMethods<T>;
    setState: (
      setter:
        | ((args: SetStateFnArgs<T>) => SetStateFnArgs<T>)
        | SetStateFnArgs<T>
    ) => void;
  }

  export interface SelectItemRenderer<T> {
    item: T;
    itemIndex?: number;
    props: SelectProps<T>;
    state: SelectState<T>;
    methods: SelectMethods<T>;
  }

  export interface SelectProps<T> {
    backspaceDelete?: boolean;
    className?: string;
    addPlaceholder?: string;
    placeholder?: string;
    loading?: boolean;
    style?: CSSProperties;
    values: T[];
    options: T[];
    multi?: boolean;
    disabled?: boolean;
    searchBy?: string;
    sortBy?: string;
    clearable?: boolean;
    searchable?: boolean;
    dropdownHandle?: boolean;
    separator?: boolean;
    keepOpen?: undefined;
    noDataLabel?: string;
    createNewLabel?: string;
    disabledLabel?: string;
    dropdownGap?: number;
    closeOnScroll?: boolean;
    debounceDelay?: number;
    labelField?: string;
    valueField?: string;
    color?: string;
    keepSelectedInList?: boolean;
    closeOnSelect?: boolean;
    clearOnBlur?: boolean;
    clearOnSelect?: boolean;
    dropdownPosition?: "top" | "bottom" | "auto";
    dropdownHeight?: string;
    autoFocus?: boolean;
    portal?: HTMLElement;
    create?: boolean;
    direction?: "ltr" | "rtl";
    name?: string;
    required?: boolean;
    pattern?: string;
    onChange: (value: T[]) => void;
    onDropdownOpen?: () => void;
    onDropdownClose?: () => void;
    onClearAll?: () => void;
    onSelectAll?: () => void;
    onCreateNew?: (item: T) => void;
    searchFn?: ({ props, state, methods }: SelectRenderer<T>) => T[];
    handleKeyDownFn?: ({
      event,
      props,
      state,
      methods,
      setState
    }: SelectKeyDown<T>) => void;
    clearRenderer?: ({
      props,
      state,
      methods
    }: SelectRenderer<T>) => JSX.Element;
    contentRenderer?: ({
      props,
      state,
      methods
    }: SelectRenderer<T>) => JSX.Element;
    dropdownRenderer?: ({
      props,
      state,
      methods
    }: SelectRenderer<T>) => JSX.Element;
    dropdownHandleRenderer?: ({
      props,
      state,
      methods
    }: SelectRenderer<T>) => JSX.Element;
    inputRenderer?: ({
      props,
      state,
      methods,
      inputRef
    }: SelectRenderer<T> & {
      inputRef: React.RefObject<HTMLInputElement>;
    }) => JSX.Element;
    itemRenderer?: ({
      item,
      itemIndex,
      props,
      state,
      methods
    }: SelectItemRenderer<T>) => JSX.Element;
    loadingRenderer?: ({ props }: SelectItemRenderer<T>) => JSX.Element;
    noDataRenderer?: ({
      props,
      state,
      methods
    }: SelectRenderer<T>) => JSX.Element;
    optionRenderer?: ({
      item,
      props,
      state,
      methods
    }: SelectItemRenderer<T>) => JSX.Element;
    separatorRenderer?: ({
      props,
      state,
      methods
    }: SelectRenderer<T>) => JSX.Element;
    additionalProps?: React.HTMLAttributes<HTMLDivElement>;
    wrapperClassName?: string;
  }

  export interface DropDownProps {
    selectBounds: DOMRect;
    dropdownGap: number;
    portal: HTMLElement;
    dropdownHeight: string;
    dropdownPosition: "auto" | "top" | "bottom";
  }

  const Select: <T extends object | string = {}>(
    props: React.PropsWithRef<SelectProps<T>>
  ) => JSX.Element;
  export default Select;
}


```

**Comment by sanusart** [comment]
Hey, thanks, why not pull request?
Looks nice but a few questions:

- What is `JSX.Element` I don't see it imported.
- Why `DropDownProps` and `DropDownState` are there but not used?
- Was it tested in real editor?

Thank you.

**Comment by 435352980** [comment]
You are wright :)
I forget to delete DropDownState and some optional declaration.
I use JSX.Element to suitable some function component width parameter type
```ts
const DropdownHandle = function<T>({ props, state, methods }: SelectRenderer<T>) {
  return <div />;
};
```
About DropDownProps
https://github.com/sanusart/react-dropdown-select/blob/cee760c1376ba684456765a8ff6dd2ef3e45e4a5/src/components/Dropdown.js#L78-L108

[About test](https://codesandbox.io/s/musing-burnell-llhlp?fontsize=14&hidenavigation=1&theme=dark) 
All types in types.d.ts

**Comment by 435352980** [comment]
Sorry, I'm a  new user of  GitHub. I don't know how to use pull request correctly.

**Comment by sanusart** [comment]
No problem at all. I just don't understand where JSX is imported from?

**Comment by 435352980** [comment]
It's a global type in @types/react
![image](https://user-images.githubusercontent.com/35485428/75425614-0b62d800-597e-11ea-958f-a0080ea51581.png)

**Comment by sanusart** [comment]
Cool, did not knew that. Thanks.

**Comment by sanusart** [comment]
Thank you! Published v4.1.1

Regarding PRs. Every time you clone a repository, make changes and try to commit them. Github will create a fork in your account and will suggest you to open a pool request.
